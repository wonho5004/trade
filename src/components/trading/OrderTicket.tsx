'use client';

import { useEffect, useState } from 'react';
import type { ChangeEvent, FormEvent } from 'react';

import { useChartStore } from '@/stores/chartStore';

const defaultForm = {
  symbol: 'BTCUSDT',
  side: 'buy',
  quantity: '0.01',
  price: ''
};

export function OrderTicket() {
  const [form, setForm] = useState(defaultForm);
  const currentSymbol = useChartStore((state) => state.symbol);

  useEffect(() => {
    setForm((prev) => ({ ...prev, symbol: currentSymbol }));
  }, [currentSymbol]);

  const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = event.target;
    setForm((prev) => ({ ...prev, [name]: value }));
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    // 주문 제출 로직은 CCXT 주문 생성 모듈 연동 후 구현 예정입니다.
    console.info('주문 요청', form);
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex w-full flex-col gap-4 rounded-lg border border-zinc-800 bg-zinc-950 p-6"
    >
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="symbol">
          거래쌍
        </label>
        <input
          id="symbol"
          name="symbol"
          value={form.symbol}
          onChange={handleChange}
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="side">
          매매 방향
        </label>
        <select
          id="side"
          name="side"
          value={form.side}
          onChange={handleChange}
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        >
          <option value="buy">매수</option>
          <option value="sell">매도</option>
        </select>
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="quantity">
          수량
        </label>
        <input
          id="quantity"
          name="quantity"
          value={form.quantity}
          onChange={handleChange}
          type="number"
          step="0.0001"
          min="0"
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="price">
          지정가 (선택)
        </label>
        <input
          id="price"
          name="price"
          value={form.price}
          onChange={handleChange}
          type="number"
          step="0.1"
          min="0"
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <button
        type="submit"
        className="rounded bg-emerald-500 px-4 py-2 text-sm font-semibold text-emerald-950 transition hover:bg-emerald-400"
      >
        주문 실행
      </button>
    </form>
  );
}
