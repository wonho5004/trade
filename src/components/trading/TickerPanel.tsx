'use client';

import { useEffect, useMemo } from 'react';
import { useShallow } from 'zustand/react/shallow';

import { useChartStore } from '@/stores/chartStore';
import { useMarketDataStore } from '@/stores/marketDataStore';

const formatNumber = (value: number | null, fractionDigits = 2) => {
  if (value === null || Number.isNaN(value)) {
    return '-';
  }
  return value.toLocaleString(undefined, {
    minimumFractionDigits: fractionDigits,
    maximumFractionDigits: fractionDigits
  });
};

export function TickerPanel() {
  const symbol = useChartStore((state) => state.symbol);
  const { status, lastPrice, priceChangePercent, highPrice, lowPrice, volume } = useMarketDataStore(
    useShallow((state) => ({
      status: state.status,
      lastPrice: state.lastPrice,
      priceChangePercent: state.priceChangePercent,
      highPrice: state.highPrice,
      lowPrice: state.lowPrice,
      volume: state.volume
    }))
  );
  const attachTicker = useMarketDataStore((state) => state.attach);

  useEffect(() => {
    if (!symbol) {
      return;
    }
    const detach = attachTicker(symbol);
    return detach;
  }, [symbol, attachTicker]);

  const isPositive = useMemo(() => (priceChangePercent ?? 0) >= 0, [priceChangePercent]);

  return (
    <section className="flex h-full flex-col justify-between rounded-lg border border-zinc-800 bg-zinc-950 p-6">
      <header className="flex items-baseline justify-between">
        <div>
          <p className="text-sm text-zinc-400">실시간 가격</p>
          <h2 className="text-lg font-semibold text-zinc-100">{symbol}</h2>
        </div>
        <span
          className={`text-sm font-semibold ${
            isPositive ? 'text-emerald-400' : 'text-rose-400'
          }`}
        >
          {status === 'connected' && priceChangePercent !== null
            ? `${formatNumber(priceChangePercent, 2)}%`
            : status === 'connecting'
            ? '연결 중...'
            : status === 'error'
            ? '에러'
            : '-'}
        </span>
      </header>

      <div className="mt-6 grid grid-cols-2 gap-4 text-sm text-zinc-400">
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">마지막 체결가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">
            {status === 'connected' ? formatNumber(lastPrice, 2) : '-'}
          </p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 고가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(highPrice, 2)}</p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 저가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(lowPrice, 2)}</p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 거래량</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(volume, 0)}</p>
        </div>
      </div>
    </section>
  );
}
