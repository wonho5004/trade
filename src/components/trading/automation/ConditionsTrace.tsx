"use client";

import { evaluateWithTrace, toExecutablePlan } from '@/lib/trading/engine/conditions';
import { buildIndicatorNumericSeries } from '@/lib/trading/engine/indicatorSignals';
import { buildActionIntents } from '@/lib/trading/engine/actions';
import type { IndicatorConditions, ConditionNode, IndicatorLeafNode } from '@/types/trading/auto-trading';
import type { EvaluationContext } from '@/lib/trading/engine/conditions';

export function ConditionsTrace({
  conditions,
  context,
  indicatorSignals,
  seriesTail
}: {
  conditions: IndicatorConditions | undefined;
  context: EvaluationContext;
  indicatorSignals?: Record<string, boolean>;
  seriesTail?: { closes: number[]; highs: number[]; lows: number[] };
}) {
  if (!conditions) return null;
  const fmt = (n: any) => {
    const v = Number(n);
    if (!Number.isFinite(v)) return String(n);
    const s = v.toFixed(4);
    return s.replace(/0+$/, '').replace(/\.$/, '');
  };
  const { result, trace } = evaluateWithTrace(conditions, context, { indicatorSignals });
  const plan = toExecutablePlan(conditions);
  const actionsPreview = (() => {
    try {
      if (!seriesTail) return [] as any[];
      const numericSeries = buildIndicatorNumericSeries(conditions, seriesTail);
      return buildActionIntents(conditions, context, indicatorSignals ?? {}, numericSeries);
    } catch {
      return [] as any[];
    }
  })();
  const indicatorMap = collectIndicatorMap((conditions.root as unknown) as ConditionNode);
  const indicatorLabel = (type: string) => {
    const map: Record<string, string> = { bollinger: '볼린저 밴드', ma: '이동평균선', rsi: 'RSI', dmi: 'DMI/ADX', macd: 'MACD' };
    return map[type] ?? type;
  };

  return (
    <div className="space-y-2 rounded border border-zinc-800 bg-zinc-950/60 p-2 text-[11px]">
      <div className="flex items-center gap-2">
        <span className="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400">세부 결과</span>
        <span className={`rounded px-2 py-0.5 ${result ? 'border border-emerald-500/60 text-emerald-200' : 'border border-zinc-700 text-zinc-300'}`}>{result ? '충족' : '불충족'}</span>
      </div>
      <div className="grid gap-2 md:grid-cols-2">
        <div>
          <p className="mb-1 text-zinc-400">지표</p>
          {plan.indicators.length === 0 ? (
            <p className="text-zinc-500">없음</p>
          ) : (
            <ul className="space-y-1">
              {plan.indicators.map((i) => (
                <li key={i.id} className="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1">
                  <span className="truncate text-zinc-300">
                    {indicatorLabel(i.type)}
                    {(() => {
                      const leaf = indicatorMap[i.id];
                      if (!leaf) return null;
                      const cmp: any = leaf.comparison;
                      if (!cmp || cmp.kind === 'none') return null;
                      if (cmp.kind === 'candle') return <span className="text-zinc-500"> · cmp candle {String(cmp.field).toUpperCase()}({cmp.reference}) {cmp.comparator}</span>;
                      if (cmp.kind === 'value') return <span className="text-zinc-500"> · cmp {cmp.comparator} {cmp.value}</span>;
                      if (cmp.kind === 'indicator') {
                        const target = indicatorMap[cmp.targetIndicatorId];
                        const tLabel = target ? indicatorLabel(target.indicator.type) : '다른 지표';
                        const metric = cmp.metric ? ` ${cmp.metric}` : '';
                        const ref = cmp.reference ? `(${cmp.reference})` : '';
                        return <span className="text-zinc-500"> · cmp {tLabel}{metric}{ref} {cmp.comparator}</span>;
                      }
                      return null;
                    })()}
                  </span>
                  <span className={`ml-2 inline-flex h-2 w-2 rounded-full ${trace[i.id] ? 'bg-emerald-400' : 'bg-zinc-600'}`} />
                </li>
              ))}
            </ul>
          )}
        </div>
        <div>
          <p className="mb-1 text-zinc-400">상태</p>
          {plan.statuses.length === 0 ? (
            <p className="text-zinc-500">없음</p>
          ) : (
            <ul className="space-y-1">
              {plan.statuses.map((s) => (
                <li key={s.id} className="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1">
                  <span className="truncate text-zinc-300">{s.metric} {s.comparator} {fmt(s.value)}{s.unit ? ` ${s.unit === 'percent' ? '%' : s.unit}` : ''}</span>
                  <span className={`ml-2 inline-flex h-2 w-2 rounded-full ${trace[s.id] ? 'bg-emerald-400' : 'bg-zinc-600'}`} />
                </li>
              ))}
            </ul>
          )}
        </div>
      </div>
      <div className="grid gap-2 md:grid-cols-2">
        <div>
          <p className="mb-1 text-zinc-400">캔들</p>
          {plan.candles.length === 0 ? (
            <p className="text-zinc-500">없음</p>
          ) : (
            <ul className="space-y-1">
              {plan.candles.map((c) => {
                const live = c.reference === 'previous' ? (context.candlePrevious as any) : (context.candleCurrent as any);
                const cur = live ? Number(live?.[c.field as any]) : null;
                return (
                  <li key={c.id} className="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1">
                    <span className="truncate text-zinc-300">
                      {String(c.field).toUpperCase()} {c.comparator} {fmt(c.value)} ({c.reference})
                      {cur != null ? <span className="text-zinc-500"> · now {fmt(cur)}</span> : null}
                    </span>
                    <span className={`ml-2 inline-flex h-2 w-2 rounded-full ${trace[c.id] ? 'bg-emerald-400' : 'bg-zinc-600'}`} />
                  </li>
                );
              })}
            </ul>
          )}
        </div>
        <div>
          <p className="mb-1 text-zinc-400">그룹</p>
          {plan.groups.length === 0 ? (
            <p className="text-zinc-500">없음</p>
          ) : (
            <ul className="space-y-1">
              {plan.groups.map((g) => (
                <li key={g.id} className="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1">
                  <span className="truncate text-zinc-300">{g.operator.toUpperCase()} · {g.id}</span>
                  <span className={`ml-2 inline-flex h-2 w-2 rounded-full ${trace[g.id] ? 'bg-emerald-400' : 'bg-zinc-600'}`} />
                </li>
              ))}
            </ul>
          )}
        </div>
      </div>
      {/* Actions preview */}
      <div>
        <p className="mb-1 text-zinc-400">액션(프리뷰)</p>
        {actionsPreview.length === 0 ? (
          <p className="text-zinc-500">없음</p>
        ) : (
          <ul className="space-y-1">
            {actionsPreview.map((a) => (
              <li key={a.id} className="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1">
                <span className="truncate text-zinc-300">
                  {a.kind.toUpperCase()} · {a.orderType || 'n/a'}
                  {a.price != null ? <span className="text-zinc-500"> · 가격 {fmt(a.price)}</span> : null}
                  {a.amount?.mode ? (
                    <span className="text-zinc-500"> · 금액 {a.amount.mode}{a.amount.value != null ? ` ${fmt(a.amount.value)}${a.amount.mode.includes('percent') ? '%' : ''}` : ''}</span>
                  ) : null}
                </span>
                <span className="text-[10px] text-zinc-500">그룹 {a.groupId}</span>
              </li>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
}

function collectIndicatorMap(root: ConditionNode): Record<string, IndicatorLeafNode> {
  const map: Record<string, IndicatorLeafNode> = {};
  const walk = (n: ConditionNode) => {
    if ((n as any).kind === 'indicator') {
      const leaf = n as IndicatorLeafNode;
      map[leaf.id] = leaf;
      return;
    }
    if ((n as any).kind === 'group') (n as any).children.forEach((c: any) => walk(c as any));
  };
  walk(root);
  return map;
}
