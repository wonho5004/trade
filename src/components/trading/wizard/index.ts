/**
 * Setup Wizard 컴포넌트 export
 */

export { SetupWizard } from './SetupWizard';
export { WelcomeStep } from './WelcomeStep';
export { ExperienceStep } from './ExperienceStep';
export { RecommendationStep } from './RecommendationStep';
export { CompleteStep } from './CompleteStep';
