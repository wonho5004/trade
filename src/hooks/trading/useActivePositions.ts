'use client';

import { useState, useEffect, useCallback } from 'react';
import type { ActivePositionSummary } from '@/types/trading/monitoring';

interface UseActivePositionsReturn {
  positions: ActivePositionSummary[];
  isLoading: boolean;
  error: string | null;
  closePosition: (id: string) => Promise<void>;
  refetch: () => Promise<void>;
}

export function useActivePositions(): UseActivePositionsReturn {
  const [positions, setPositions] = useState<ActivePositionSummary[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  const fetchPositions = useCallback(async () => {
    setIsLoading(true);
    setError(null);

    try {
      const response = await fetch('/api/monitoring/positions');
      const data = await response.json();

      if (!response.ok) {
        // 401 에러는 조용히 처리 (로그인 필요)
        if (response.status === 401) {
          setError(null); // 에러 표시 안함
          setIsLoading(false);
          return;
        }
        throw new Error(data.error?.message || '포지션 조회에 실패했습니다');
      }

      setPositions(data.positions || []);
    } catch (err) {
      const message = err instanceof Error ? err.message : '알 수 없는 오류가 발생했습니다';
      setError(message);
      console.error('Failed to fetch positions:', err);
    } finally {
      setIsLoading(false);
    }
  }, []);

  const closePosition = useCallback(async (id: string) => {
    const response = await fetch(`/api/monitoring/positions/${id}/close`, {
      method: 'POST'
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.error?.message || '포지션 청산에 실패했습니다');
    }

    // Refetch to update the list
    await fetchPositions();
  }, [fetchPositions]);

  // Initial fetch and periodic updates
  useEffect(() => {
    fetchPositions();

    // Poll every 5 seconds for real-time updates
    const interval = setInterval(fetchPositions, 5000);

    return () => clearInterval(interval);
  }, [fetchPositions]);

  return {
    positions,
    isLoading,
    error,
    closePosition,
    refetch: fetchPositions
  };
}
