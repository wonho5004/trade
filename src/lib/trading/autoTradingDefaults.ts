import type {
  AutoTradingSettings,
  AdditionalEntryConditions,
  CapitalSettings,
  EntryConditions,
  EntryDirectionSettings,
  ExitConditions,
  ExitDirectionSettings,
  HedgeActivationSettings,
  IndicatorConditions,
  IndicatorConfigMap,
  IndicatorEntry,
  IndicatorKey,
  AggregatorOperator,
  ConditionGroupNode,
  ConditionNode,
  IndicatorLeafNode,
  CandleLeafNode,
  InitialMarginSetting,
  PositionDirection,
  ProfitCondition,
  ScaleInBudget,
  ScaleInDirectionSettings,
  ScaleInLimit,
  StopLossConditions,
  StopLossLineSettings,
  SymbolSelection,
  ThresholdCondition,
  IndicatorComparisonLeaf,
  CandleCondition,
  ComparisonOperator,
  CandleFieldOption,
  CandleReference
} from '@/types/trading/auto-trading';

export const createProfitCondition = (): ProfitCondition => ({
  enabled: false,
  comparator: 'over',
  value: 0
});

export const createThresholdCondition = (): ThresholdCondition => ({
  enabled: false,
  comparator: 'over',
  value: 0
});

const createIndicatorEntryId = () => `ind-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;

const createConditionNodeId = () => `cond-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;

export const createCandleCondition = (): CandleCondition => ({
  enabled: false,
  field: 'close',
  comparator: 'over',
  targetValue: 0,
  reference: 'current'
});

export const createCandleLeaf = (candle?: CandleCondition): CandleLeafNode => ({
  kind: 'candle',
  id: createConditionNodeId(),
  candle: candle ? { ...candle } : createCandleCondition()
});

export const createBollingerCondition = (): IndicatorConfigMap['bollinger'] => ({
  enabled: false,
  length: 20,
  standardDeviation: 2,
  offset: 0,
  band: 'middle',
  action: 'touch',
  touchTolerancePct: 0.2
});

export const createMaCondition = (): IndicatorConfigMap['ma'] => ({
  enabled: false,
  period: 20,
  actions: []
});

export const createRsiCondition = (): IndicatorConfigMap['rsi'] => ({
  enabled: false,
  period: 14,
  smoothing: 'sma',
  threshold: 50,
  actions: []
});

export const createDmiCondition = (): IndicatorConfigMap['dmi'] => ({
  enabled: false,
  diPeriod: 14,
  adxPeriod: 14,
  adx: createThresholdCondition(),
  diComparison: null,
  diPlus: createThresholdCondition(),
  diMinus: createThresholdCondition(),
  adxVsDi: null
});

export const createMacdCondition = (): IndicatorConfigMap['macd'] => ({
  enabled: false,
  fast: 12,
  slow: 26,
  signal: 9,
  source: 'close',
  method: 'EMA',
  comparison: null,
  histogramAction: null
});

export const createIndicatorConfig = <T extends IndicatorKey>(indicator: T): IndicatorConfigMap[T] => {
  switch (indicator) {
    case 'bollinger':
      return createBollingerCondition() as IndicatorConfigMap[T];
    case 'ma':
      return createMaCondition() as IndicatorConfigMap[T];
    case 'rsi':
      return createRsiCondition() as IndicatorConfigMap[T];
    case 'dmi':
      return createDmiCondition() as IndicatorConfigMap[T];
    case 'macd':
      return createMacdCondition() as IndicatorConfigMap[T];
    default:
      return createBollingerCondition() as IndicatorConfigMap[T];
  }
};

export const createIndicatorEntry = <T extends IndicatorKey>(indicator: T): IndicatorEntry<T> => ({
  id: createIndicatorEntryId(),
  type: indicator,
  config: createIndicatorConfig(indicator)
});

export const createIndicatorComparisonLeaf = (
  legacy: unknown = { kind: 'none' },
  fallbackId: string
): IndicatorComparisonLeaf | { kind: 'none' } => {
  if (!legacy || typeof legacy !== 'object') {
    return { kind: 'none' };
  }
  const candidate = legacy as any;
  if (candidate.kind === 'candle' && candidate.field && candidate.reference && candidate.comparator) {
    return {
      kind: 'candle',
      comparator: candidate.comparator,
      field: candidate.field,
      reference: candidate.reference
    };
  }
  if (candidate.kind === 'value' && typeof candidate.value === 'number' && candidate.comparator) {
    return {
      kind: 'value',
      comparator: candidate.comparator,
      value: candidate.value
    };
  }
  if (candidate.kind === 'indicator' && candidate.comparator) {
    return {
      kind: 'indicator',
      comparator: candidate.comparator,
      targetIndicatorId: candidate.targetIndicatorId ?? fallbackId
    };
  }
  return { kind: 'none' };
};

export const createIndicatorLeaf = (
  indicator: IndicatorEntry,
  comparison: IndicatorComparisonLeaf | { kind: 'none' } = { kind: 'none' }
): IndicatorLeafNode => ({
  kind: 'indicator',
  id: createConditionNodeId(),
  indicator,
  comparison
});

export const createIndicatorGroup = (
  operator: AggregatorOperator = 'and',
  children: ConditionNode[] = []
): ConditionGroupNode => ({
  kind: 'group',
  id: createConditionNodeId(),
  operator,
  children
});

export const createIndicatorConditions = (operator: AggregatorOperator = 'or'): IndicatorConditions => {
  // Stable default tree with deterministic IDs for test equality
  const candle = createCandleCondition();
  const root: ConditionGroupNode = {
    kind: 'group',
    id: 'cond-root',
    operator,
    children: []
  };

  // Return a legacy-compatible shape for transitional tests
  return {
    root,
    candle,
    entries: [],
    defaultAggregator: root.operator
  } as unknown as IndicatorConditions;
};

export const normalizeConditionTree = (root: ConditionNode): ConditionGroupNode => {
  // Defensive: tolerate unexpected/legacy shapes where `root` may be missing or malformed
  const isValidNode = (n: any) => n && typeof n === 'object' && typeof n.kind === 'string';
  const base: ConditionNode = isValidNode(root)
    ? (root as ConditionNode)
    : createIndicatorGroup('and', []);

  const collectIndicatorIds = (node: ConditionNode, acc: Set<string>) => {
    if (node.kind === 'indicator') {
      acc.add(node.id);
      return;
    }
    if (node.kind === 'group') {
      node.children.forEach((child) => collectIndicatorIds(child, acc));
    }
  };

  const sanitizeComparisons = (node: ConditionNode, indicatorIds: Set<string>): ConditionNode => {
    if (node.kind === 'indicator') {
      if (node.comparison.kind === 'indicator' && !indicatorIds.has(node.comparison.targetIndicatorId)) {
        return { ...node, comparison: { kind: 'none' } };
      }
      return node;
    }
    if (node.kind === 'group') {
      return {
        ...node,
        children: node.children.map((child) => sanitizeComparisons(child, indicatorIds))
      };
    }
    return node;
  };

  const ensureGroupRoot = (node: ConditionNode): ConditionGroupNode => {
    if (node.kind === 'group') return { ...node };
    return createIndicatorGroup('and', [node]);
  };

  const indicatorIds = new Set<string>();
  collectIndicatorIds(base, indicatorIds);
  const sanitized = sanitizeComparisons(base, indicatorIds);
  return ensureGroupRoot(sanitized);
};

const createInitialMarginSetting = (): InitialMarginSetting => ({
  basis: 'wallet',
  mode: 'per_symbol_percentage',
  percentage: 5,
  minNotional: 0,
  usdtAmount: 0,
  asset: 'USDT'
});

const createScaleInBudget = (): ScaleInBudget => ({
  mode: 'balance_percentage',
  basis: 'wallet',
  percentage: 5,
  minNotional: 0,
  usdtAmount: 0,
  asset: 'USDT'
});

const createScaleInLimit = (): ScaleInLimit => ({
  balance: {
    enabled: false,
    basis: 'wallet',
    percentage: 10
  },
  initialMargin: {
    enabled: false,
    percentage: 100
  },
  purchaseCount: {
    enabled: false,
    comparator: 'over',
    value: 1
  },
  profitRate: createProfitCondition(),
  unlimited: true
});

export const createCapitalSettings = (): CapitalSettings => ({
  estimatedBalance: 10000,
  maxMargin: {
    basis: 'wallet',
    percentage: 20
  },
  initialMargin: createInitialMarginSetting(),
  scaleInBudget: createScaleInBudget(),
  scaleInLimit: createScaleInLimit(),
  useMinNotionalFallback: true,
  hedgeBudget: {
    separateByDirection: false,
    long: { mode: 'position_percent', percentage: 100 },
    short: { mode: 'position_percent', percentage: 100 }
  }
});

export const createEntryDirectionSettings = (direction: PositionDirection): EntryDirectionSettings => ({
  enabled: false,
  immediate: false,
  indicators: createIndicatorConditions()
});

export const createEntryConditions = (): EntryConditions => ({
  long: createEntryDirectionSettings('long'),
  short: createEntryDirectionSettings('short')
});

export const createScaleInDirectionSettings = (direction: PositionDirection): ScaleInDirectionSettings => ({
  enabled: direction === 'long',
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions(),
  currentBuyNotional: { enabled: false, asset: 'USDT', comparator: 'none', value: 0 },
  initialBuyChange: { enabled: false, comparator: 'none', value: 0 }
});

export const createScaleInConditions = (): AdditionalEntryConditions => ({
  long: createScaleInDirectionSettings('long'),
  short: createScaleInDirectionSettings('short')
});

export const createExitDirectionSettings = (direction: PositionDirection): ExitDirectionSettings => ({
  enabled: direction === 'long',
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions(),
  includeFeesFunding: false,
  currentBuyNotional: { enabled: false, asset: 'USDT', comparator: 'none', value: 0 },
  initialBuyChange: { enabled: false, comparator: 'none', value: 0 },
  sellAmount: { mode: 'percent', value: 0 }
});

export const createExitConditions = (): ExitConditions => ({
  long: createExitDirectionSettings('long'),
  short: createExitDirectionSettings('short')
});

export const createStopLossLineSettings = (): StopLossLineSettings => ({
  enabled: false,
  autoRecreate: true,
  profitTarget: createProfitCondition(),
  indicators: createIndicatorConditions()
});

export const createStopLossConditions = (): StopLossConditions => ({
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  purchaseCount: {
    enabled: false,
    comparator: 'over',
    value: 1
  },
  stopLossLine: createStopLossLineSettings(),
  indicators: createIndicatorConditions()
});

export const createSymbolSelection = (): SymbolSelection => ({
  manualSymbols: [],
  ranking: {
    market_cap: null,
    volume: null,
    top_gainers: null,
    top_losers: null
  },
  excludedSymbols: [],
  excludedReasons: {},
  respectDefaultExclusions: true,
  leverageMode: 'uniform',
  leverageOverrides: {},
  positionOverrides: {},
  featureOverrides: {},
  maxListingAgeDays: null,
  rankingSort: 'alphabet',
  autoFillRecheck: false,
  excludeTopGainers: null,
  excludeTopLosers: null,
  excludeBottomVolume: null,
  excludeBottomMarketCap: null
});

export const createHedgeActivationSettings = (): HedgeActivationSettings => ({
  enabled: false,
  directions: ['long', 'short'],
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions()
});

export const createDefaultAutoTradingSettings = (): AutoTradingSettings => ({
  logicName: '',
  leverage: 5,
  symbolCount: 1,
  assetMode: 'single',
  positionMode: 'one_way',
  capital: createCapitalSettings(),
  timeframe: '3m',
  symbolSelection: createSymbolSelection(),
  entry: createEntryConditions(),
  scaleIn: createScaleInConditions(),
  exit: createExitConditions(),
  stopLoss: createStopLossConditions(),
  hedgeActivation: createHedgeActivationSettings(),
  metadata: {
    lastSavedAt: null,
    lastValidatedAt: null
  }
});
