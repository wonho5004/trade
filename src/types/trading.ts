export type OrderSide = 'buy' | 'sell';

export interface ExchangeCredentials {
  apiKey: string;
  secret: string;
  password?: string;
}

export interface OrderRequest {
  symbol: string;
  side: OrderSide;
  quantity: number;
  price?: number;
  type: 'market' | 'limit';
}
