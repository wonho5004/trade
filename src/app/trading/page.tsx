import { AppHeader } from '@/components/common/AppHeader';
import { OrderTicket } from '@/components/trading/OrderTicket';
import { TickerPanel } from '@/components/trading/TickerPanel';

export default function TradingPage() {
  return (
    <div className="flex min-h-screen flex-col gap-6 bg-zinc-950 text-zinc-100">
      <AppHeader />
      <main className="flex flex-1 flex-col gap-6 px-6 pb-10">
        <section className="grid gap-6 md:grid-cols-[3fr,2fr]">
          <TickerPanel />
          <OrderTicket />
        </section>
      </main>
    </div>
  );
}
