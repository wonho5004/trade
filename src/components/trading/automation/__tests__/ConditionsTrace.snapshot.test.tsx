// @ts-nocheck
import React from 'react';
import { render } from '@testing-library/react';

import { ConditionsTrace } from '../ConditionsTrace';
import { createIndicatorEntry, createIndicatorLeaf, createIndicatorGroup } from '@/lib/trading/autoTradingDefaults';

function sampleConditions() {
  const ma = createIndicatorEntry('ma');
  (ma.config as any).period = 20;
  const leaf = createIndicatorLeaf(ma, { kind: 'value', comparator: 'over', value: 0 });
  const root = createIndicatorGroup('and', [leaf]);
  (root as any).id = 'cond-root';
  return { root } as any;
}

describe('ConditionsTrace snapshots', () => {
  it('renders indicator/status/candle summary consistently', () => {
    const conds = sampleConditions();
    const context: any = {
      symbol: 'BTCUSDT',
      direction: 'long',
      candleCurrent: { timestamp: Date.now(), open: 1, high: 2, low: 0.5, close: 1.5, volume: 1000 },
      candlePrevious: { timestamp: Date.now() - 60000, open: 0.9, high: 1.1, low: 0.8, close: 1.0, volume: 900 }
    };
    const indicatorSignals = { /* by node id */ } as any;
    const { container } = render(<ConditionsTrace conditions={conds} context={context} indicatorSignals={indicatorSignals} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
