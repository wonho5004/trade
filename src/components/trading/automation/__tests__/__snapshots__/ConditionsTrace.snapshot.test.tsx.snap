// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionsTrace snapshots renders indicator/status/candle summary consistently 1`] = `
<div
  class="space-y-2 rounded border border-zinc-800 bg-zinc-950/60 p-2 text-[11px]"
>
  <div
    class="flex items-center gap-2"
  >
    <span
      class="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400"
    >
      세부 결과
    </span>
    <span
      class="rounded px-2 py-0.5 border border-zinc-700 text-zinc-300"
    >
      불충족
    </span>
  </div>
  <div
    class="grid gap-2 md:grid-cols-2"
  >
    <div>
      <p
        class="mb-1 text-zinc-400"
      >
        지표
      </p>
      <ul
        class="space-y-1"
      >
        <li
          class="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1"
        >
          <span
            class="truncate text-zinc-300"
          >
            이동평균선
            <span
              class="text-zinc-500"
            >
               · cmp 
              over
               
              0
            </span>
          </span>
          <span
            class="ml-2 inline-flex h-2 w-2 rounded-full bg-zinc-600"
          />
        </li>
      </ul>
    </div>
    <div>
      <p
        class="mb-1 text-zinc-400"
      >
        상태
      </p>
      <p
        class="text-zinc-500"
      >
        없음
      </p>
    </div>
  </div>
  <div
    class="grid gap-2 md:grid-cols-2"
  >
    <div>
      <p
        class="mb-1 text-zinc-400"
      >
        캔들
      </p>
      <p
        class="text-zinc-500"
      >
        없음
      </p>
    </div>
    <div>
      <p
        class="mb-1 text-zinc-400"
      >
        그룹
      </p>
      <ul
        class="space-y-1"
      >
        <li
          class="flex items-center justify-between rounded border border-zinc-800 bg-zinc-950/40 px-2 py-1"
        >
          <span
            class="truncate text-zinc-300"
          >
            AND
             · 
            cond-root
          </span>
          <span
            class="ml-2 inline-flex h-2 w-2 rounded-full bg-zinc-600"
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;
